Composite compliance for index_copy, index_fill, masked_scatter, masked_fill

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71766
* #71765
* __->__ #71751
* #71750

Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I deleted the autograd formula for the in-place variant and replaced
it with the out-of-place variant
- This makes the forward-ad formula slightly slower because the codegen
generates a temporary but offline discussion suggests it's not worth
maintaining two sets of formulas for this and we can make the autograd
codegen smarter in the future.
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

Differential Revision: [D33772756](https://our.internmc.facebook.com/intern/diff/D33772756)

zou3519 committed:
Composite compliance for index_copy, index_fill, masked_scatter, masked_fill

Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I duplicated the autograd formula with the in-place variant. This step
is already required to add a structured kernel and you can view this PR
as taking one step towards adding structured kernels for these ops. For
performance reasons, I cannot delete the autograd formula for the
in-place variant (but I am more than happy to if you want me to).
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

[ghstack-poisoned]

zou3519 commented:
XLA tests are failing, but I think that's because the lowerings are defined in terms of the in-place operations (and not the out-of-place operations). If we agree that this is the right way forward then I'll open an issue on the pytorch/xla side.

zou3519 committed:
Update on "Composite compliance for index_copy, index_fill, masked_scatter, masked_fill"


Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I deleted the autograd formula for the in-place variant and replaced
it with the out-of-place variant
- This makes the forward-ad formula slightly slower because the codegen
generates a temporary but offline discussion suggests it's not worth
maintaining two sets of formulas for this and we can make the autograd
codegen smarter in the future.
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33772756).

albanD reviewed (commented):
Thanks for the update! Looks good.

cc @soulitzer we should make the forward AD codegen smarter for such cases and generate `self_t.foo_()` from `self_t.foo()` instead of `self_t.copy_(self_t.foo())`.

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5821,6 +5823,8 @@
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CPU, CUDA: masked_fill
```
Is there a reason why these can't be `CompositeExplicitAutograd` instead?

I think that's why the XLA tests are failing in https://github.com/pytorch/xla/issues/3314: `masked_fill` is no longer composite, so calling `at::masked_fill()` with an xla tensor will dispatch directly to the `XLA` key. There's no kernel registered directly to `masked_fill()`, so we hit the CPU fallback, causing the test to fail.

If it were `CompositeExplicitAutograd` instead, then XLA would end up re-using the `masked_fill` kernel in core, which redispatches to `masked_fill_()` (which XLA has a lowering for).

zou3519 replied:
My understanding is that CompositeExplicitAutograd operations should work for all/most backends (please correct me if that's not supposed to be the case).

I didn't make this CompositeExplicitAutograd because its implementation is `self.clone().masked_fill_(source)`.  masked_fill_ only has CPU, CUDA dispatch, so masked_fill only should support the CPU and CUDA backends.

That being said I don't know how XLA works and am happy to change this to CompositeExplicitAutograd if we think it's the right thing to do.

bdhirsh replied:
You're right that since the implementation is `self.clone().masked_fill_(source)`, then even though it's "composite", a backend will only support it if it provides kernels for those ops (`clone` and `masked_fill_`).

In this case though, that's pretty much what the current state is today, since `masked_fill()` is already a `CompositeImplicitAutograd` op in master. And XLA [currently has an explicit lowering](https://github.com/pytorch/xla/blob/343b475ffd2a3c505531dd93a05a4cdf924c0ed3/xla_native_functions.yaml#L176) for `masked_fill_()`, basically for that reason.

So I think that we should make these kernels all `CompositeExplicitAutograd` for now to preserve the current state. The better long term state will be for these ops to all become structured (and XLA will be able to just directly lower the out-of-place `masked_fill()` instead.

albanD replied:
I am not sure that is the idea.
If a backend wants to override this op, then if we specify the keys for both, they have to re-implement both functions.
But if masked_fill is compositeexplicit. Then they only have to implement masked_fill_ to get both working.
So I would agree with @bdhirsh here.

zou3519 replied:
Cool, thanks for the discussion, I will change to CompositeExplicitAutograd.

zou3519 replied:
Question: does changing it to CompositeExplicitAutograd prevent XLA from registering a lowering for e.g. masked_fill if they desired? For a in-tree backend I expect this would be possible by specifying the different dispatch for different backend.

bdhirsh replied:
Nope - Registering a kernel as `CompositeExplicitAutograd` means that every backend will get that kernel by default, but can override with their own kernel if they want.

XLA also has a boxed fallback for all ops that falls back to CPU. But that gets an even lower precedence than the composite kernel. The full precedence order defined in the dispatcher registration logic is:
(1) direct kernel registrations
(2) composite registrations (if no direct registration defined)
(3) boxed fallback (if 1 and 2 both aren't available)

zou3519 replied:
Awesome, thanks for the detailed explanation!

albanD commented:
nit: opened an issue for the perf regression in fw ad: https://github.com/pytorch/pytorch/issues/71987

zou3519 committed:
Update on "Composite compliance for index_copy, index_fill, masked_scatter, masked_fill"


Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I deleted the autograd formula for the in-place variant and replaced
it with the out-of-place variant
- This makes the forward-ad formula slightly slower because the codegen
generates a temporary but offline discussion suggests it's not worth
maintaining two sets of formulas for this and we can make the autograd
codegen smarter in the future.
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

Differential Revision: [D33772756](https://our.internmc.facebook.com/intern/diff/D33772756)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33772756).

zou3519 committed:
Update on "Composite compliance for index_copy, index_fill, masked_scatter, masked_fill"


Before this PR, each of the above operations were composite and had
in-place variants that were primitive w.r.t. autograd.

The backward passes are not composite compliant due to the op (e.g.
index_copy) decomposing into index_copy_ and index_copy_'s backward
formula having in-place operations in it. To fix this, for each of the
ops mentioned in the title:
- I deleted the autograd formula for the in-place variant and replaced
it with the out-of-place variant
- This makes the forward-ad formula slightly slower because the codegen
generates a temporary but offline discussion suggests it's not worth
maintaining two sets of formulas for this and we can make the autograd
codegen smarter in the future.
- I then replaced instances of grad.clone().inplace_variant_(...) with
grad.outplace_variant(...)

Test Plan:
- run existing tests to check correctness
- run functorch tests

Differential Revision: [D33772756](https://our.internmc.facebook.com/intern/diff/D33772756)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33772756).
